src/mpyl/utilities/docker/__init__.py

```diff
@@ -144,6 +158,16 @@ def docker_registry_path(docker_config: DockerConfig, image_name: str) -> str:
     return "/".join([c for c in path_components if c]).lower()
 
 
+def registry_for_project(
+    docker_config: DockerConfig, project_docker: Optional[Docker]
+) -> DockerRegistryConfig:
+    if project_docker:
+        for registry in docker_config.registries:
+            if registry.host_name == project_docker.registry_url:
+                return registry
+    return docker_config.registries[0]
```

Comment:
Should we allow for a fallback or should be actually throw an error here (you know better than me)?

Reply:
Most `project.yml` have no `docker` block so think we still need default behavior. Either taking the first one or having one marked as default explicitly?

Reply:
Why don't we define custom docker registries in `project.yml` only? That seems more in line with the way we've been using `mpyl_config.yml` so far: to provide global settings and `project.yml` defaults.

Reply:
Also, currently the `docker` field in the project yaml has the following structure:
```
docker:
  credentialsId: 'dockerExternalRegistry'
  registryUrl: 'vdbexternalaccessregistry.azurecr.io'
```
I think we need to change the schema and the mechanism here, because we're referring to a jenkins credential here, which I don't think we have access to in mpyl

Reply:
Right, I think I understand the problem better now.
The key thing I misunderstood is that this is configuration that is needed during the _build_ stage. I see two ways to solve this:
1. _Turn de docker registry config into a list as @peterrutg proposed and reference it from `project.yml`_
   - *Pro* avoids credential config duplication of `project.yml`
   - *Pro* allows us to inject build time credentials in a way that we're already used to.
   - *Con* introduces reference from `project.yml` to `mpyl_conig.yml` that is not immediately obvious to the reader and cannot be validated by the schema.
   - *Con* introduces breaking change to the schema. 
2. _Define a single default docker registry in the `mpyl_config.yml` and allow this to be overridden at `project.yml` level._
   - *Pro* quite explicit and which projects will get non-default values. Especially in scenarios where we have many (say > 10) registries used by only specific projects, defining them in the `project.yml` would be explicit and reduce the chances of stale configuration in `mpyl_config.yml` that nobody knows whether it's still in use.
   - *Con* no possibility to do build time env substitution of `project.yml`. This shouldn't be a difficult feature to add, but it makes the project config less declarative. It may not be obvious to the user when the variables will be substituted. And perhaps also open the door to "powerful" or "clever" possibilities that I can't foresee at the moment (project.yml override comes to mind).  

I'm personally tending towards option 1, with a few implementation notes:
 - would benefit from automated upgrade script for `mpyl_config.yml` as introduced by this PR https://github.com/Vandebron/mpyl/pull/253/commits/0612f26b1c5881e7b74b48ac2cea501ab6f44a3c
 - I would suggest to introduce a default registry field to avoid having to define the registry to use in all `project.yml`. Something like this:
 ```
 docker:
  defaultRegistry: 'bigdataregistry.azurecr.io'
  registries:
    - hostName: 'bigdataregistry.azurecr.io'
      userName: !ENV ${DOCKER_REGISTRY_USR:docker_user}
      password: !ENV ${DOCKER_REGISTRY_PSW:docker_password}
    - hostName: 'vdbexternalaccessregistry.azurecr.io'
      userName: !ENV ${DOCKER_EXTERNAL_REGISTRY_USR:docker_user}
      password: !ENV ${DOCKER_EXTERNAL_REGISTRY_PSW:docker_password}
 ```
  - The `credentialsId` and `registryUrl` fields aren't used by MPyL so I think we can deprecate and remove them.
  - What about the `imagePullSecrets` that you need deploy time? I guess they relate to the registry that you've chosen to push your images too. Although that doesn't _have_ to be the case, in case we want to use images built by an external party.
  
@koeves @peterrutg What do you think?

Reply:
https://github.com/Vandebron/mpyl/pull/253 with config upgrade mechanism is ready for review.

Reply:
@peterrutg the config upgrade is in master. I've rebased this branch.

Reply:
Cool, thanks!